### PR TITLE
Feature/autoruncontrol

### PR DESCRIPTION
**Pull request by momoadeli**
**_(replaces previous pull request.  removed majority of comments --left a couple of suggested todo's--and also removed git references to #9 (altho you seemed to have merged) . Also, screenshots below.)_**

Kirill, here is a draft of a working autorun ON/OFF functionality. Please feedback before I move further (cleanup, redundancies, etc)--I want to make sure I leave this in a state that you above all are comfortable with for obvious reasons. BTW, there are no 3d party libs (removed bootstrap). A few quick items:

1. There are a few ways autorun ON/OFF could be implemented. Currently I have a 'decoupled' autoruncontrol.component.ts) which @Outputs its 'state' via OnAutorunChange to its parent exercise.component.ts.. After your feedback, instead of the emitted (@Output), I would like to enhance this where the parent can pass through an (@Input) method to this component to use for 'run-click' which I believe is a cleaner pattern.
1A: Needless to say, the autorunner has a 'toggler' The proper[ActionType.TOGGLE_AUTORUN] method has been added to the reducers.service.ts.
2. exercise.component.ts: Here I've added a (@Output)OnAutorunChange which is triggered by autorun.component.ts similar to (@Output)onCodeChange which is triggered by its editors.components.ts child. Both, one way or another, eventually, execute a `this.state.updateCode(changedFile`) which is handled by the `state.service.ts`.
3. I've tried to follow your current styles and patterns with the other knick-knack code I've added to build this here and there.
4. I'm not a UI designer, so let me know if you're comfortable with the autoruncontrol.component.ts look and feel. There are different ways this can be done too.

Looking forward to feedback. Talk soon!

screen shot 2016-12-05 at 12 15 02 pm
screen shot 2016-12-05 at 12 19 17 pm

screen shot 2016-12-05 at 12 15 20 pm

### CODE REVIEW

File: src/app/state.service.ts [hunk 1]
```diff
@@ -82,6 +82,14 @@ export class StateService {
     this.dispatchAction(ActionTypes.NEXT_EXERCISE);
   }
 
+  toggleAutorun() {
+    this.dispatchAction(ActionTypes.TOGGLE_AUTORUN);
+  }
+
+  getAutorun() {
```
Comment (kirjs) [inline src/app/state.service.ts]: I don't see this used

**Comment by kirjs** [comment]
Screenshots looks nice. 

I went through the the code, and have a bunch of (mostly minor) suggestions. 
Depending on the amount of time you have I could either do a thorough code review, or just merge it in and fix things myself (both would probably take me equal amount of time).
Let me know!

Also I'm changing a bunch of things again (this time I'm actually almost happy with where I'm getting to, so hopefully this will  be the last time), so let me know if you need help merging this in.

File: src/app/editors/editors.component.ts [hunk 2]
```diff
@@ -2,6 +2,8 @@ import {Component, Input, Output, EventEmitter} from '@angular/core';
 import {Action} from "../action";
 import {FileConfig} from "../file-config";
 import {StateService} from "../state.service";
+//momo
+import { AutorunControlInterface } from "../exercise/autoruncontrol.interface"
```
Comment (kirjs) [inline src/app/editors/editors.component.ts]: Don't see this used.

File: src/app/state.service.ts (hunk 1 above)
Comment (momoadeli) [reply src/app/state.service.ts]: @kirjs , I removed it. It seemed, initially `state.service` to be the best place to maintain state for that but it's behind by a `onCodeChange` cycle. Currently, exercise.component maintains that autorun state. This was the was driving reason to, perhaps, have a simple autoruncontrol.service for that. Thoughts welcome.

File: src/app/editors/editors.component.ts (hunk 2 above)
Comment (momoadeli) [reply src/app/editors/editors.component.ts]: @kirjs , removed.  Note the interface has a 'running' attribute that might go away if one of the TODO's--passing in an (@Input)run() method from `exercise,component ` sits well with you.  Thoughts welcome.

**Comment by momoadeli** [comment]
@kirjs ,please feel free to merge mine in (I might slow you down otherwise). I'll update my fork accordingly (although still not sure why I'm seeing my fork behind your repo already--ergo the previous 'spec' code I removed from my branch). I'm going to continue with improvements. Note, there are things that should be added (eg: disable/enable run button on code execution, or properly manage autorun toggle conflict in the midst of 'run click'. I submitted this PR so I don't digress too far from what you like before getting feedback)

Meanwhile, I'll scan the other issues to see if I can add a hand if that's ok with you...:)

**Comment by kirjs** [comment]
Thank you! Will merge it in later today!
For sure, feel free to contribute more if you find something interesting

**Comment by kirjs** [comment]
Just got this in, actually I changed things a bit to have a `RUN_CODE` reducer instead of diffing the code in the runner component like I did before, and this improved the speed a lot. 

Thanks for the nice start!

**Pull request merged**